README.md
# RadarTracker
A Factorio mod, that adds "trackers", which can track **trains, cars, tanks (any vehicle-type) and immoveables like train-stops and roboports**..

Graphics: Trackers by YuokiTani. Thanks to him.

# Description
Radar-tracker tries to let you keep track of your factory: It triggers _charting_ (actualization of the map) only for the "most important" things.
- The Movement-tracker tracks *vehicles* (trains, cars, tanks):
  - Technique can be researched with Logistics-2.
  - Moving vehicles will be charted every 0.5 seconds, and the radars scans into the forward-direction.
  - Waiting vehicles will be charted every 17.5 seconds.
- The Immoveables-tracker tracks train-stops and roboports every 30 seconds.

---------------------------------------

# Long description
This mod is for bigger factories, that already covers a wide area and a bigger train-network.
But it is also useful for exploration by car/tank.

## About Charting (technical) and why it is bad for game-performance
Charting is the action, when a radar scans a chunk. Only whole chunks (32x32 tiles) can be charted.

Once a chunk is charted, it keeps up-to-date for **10 seconds** (this is default game-mechanics).
Within that time the piece of map for this chunk is updated for every tick.
You can see that cleary when you look on the map and a chunk suddenly is highlighted and then gets darker and darker until it is the same color as the other uncharted chunks.
In other words: 10 seconds after the last chart of a chunk the map of this chunk falls back into the "fog of war": The map is not re-charted and what you see on map might be outdated.

This means: The default radar charts an area of 100x100 tiles every tick (100x100 means the chunks,
that are "touched" by this area will be charted).
 
**This is an CPU-expensive operation if you place many radars (many hundreds of chunks)!**

Much more expensive are radars (from other mods), that chart a wider area: An area of 200x200 is 4 times larger, than 100x100. Not 2.

So for very big factories it is a good idea to reduce the number of radars. But with growing factory you just need the overview!

## What you can do with it?

- See always all your trains.
- Leave a car in your outpost and see (up to) 4 chunks every 17.5 seconds around the car.
- See, what's going on at the train stations (every 30 seconds).
- Drive around with your car and see longer scan-range in driving-direction.

## Goals
This mod tries to achieve this goals:
- Remove stupid game-play to place more and more radars to keep track.
- Keep track of your *most important* stuff without always need to place another radar.
- Reduce need for CPU-expensive full-scan radar and replace with a "good-enough" charting.
- Add a new game-play that it makes sense to switch the new trackers off if they take too much energy.
- Radar only for intensive non-stop-scanning.

---------------------------------------

# Docs

## Entities
You can research the entities in the following order:

### Movement tracker
Charts every vehicle (trains, cars, tank...).

If vehicle is driving: Every 1/2 seconds and tries to look forward. (This is with the default vehicles good enough, but there might be problems with very fast vehicles)

If it is waiting: Every 17.5 seconds (10 sconds visible, 7.5 in fog).

### Immovebles tracker
Charts train-stops and roboports every 30 seconds. No need to place radars everywhere.

This means: Train-stops (and roboports) are charted once every 30 seconds; they keep charted 10 seconds completly and 20 seconds in fog-of-war (until it is re-charted). This is with default trains a good compromise.

### Rotation tracker

**THE ROTATION-TRACKER IS NOT READY YET (and might change name); you can research and build it, but the entity has currently no function!**
When ready it should scan a very wide range (2000 tiles?) in "circle-segments", like real radars. Only chunks with owned entities on it.
Takes some time for a rotation (a minute or two).

### Research
You need to research the different radar-tracker-technologies before you can built and place the radars.
TODO: List needed items for research

### Crafting
TODO: List of needed items to built

### Energy Usage
TODO: (auto-generated) list of needed power

# Remote Calls / Debug
Use /c remote.call('RT', 'help') to see all commands.
TODO: Generate list

---------------------------------------

# Features:
- This is my first Factorio mod, I appreciate positive and negative critics.
- Fast initialization: Initial scan of a 20,000 tiles-map in under 10 seconds (vs. half an hour). 
- Tracks only first and last unit of a train instead of the whole train. (Improvement is planned, see ideas)
- CPU-save: For 300 trains (600 entities) only 3-6 updates per second slower. (Well, still too slow)
- Debug-mode. Useful and interesting...
- Configuration file for energy-usages etc.

---------------------------------------

#Background
To be honest, the biggest reason why I did this was, that I needed something to see my trains on my 20km² map.
I had overseen, that Optera did a 0.13 fix, so I fixed it too for myself (cause I needed it for playing on my really big map).
I think I wouldn't have made it, if I had read the whole thread. But when I was on it I fixed here a bit and there a bit
and I had some more ideas (The second reason was learning more Lua programming and Factorio mods of course).
And suddenly it was as it was. Then I though: Would be a pity not to publish it. Forgive me, it's my first mod. :)

## Originating from Train-Tracker
This mod is is originated from **Train-Tracker** by MrDoomah.
https://forums.factorio.com/viewtopic.php?f=92&t=18368
It was updated to v0.13 by Optera (in the same thread) and me.

Meanwhile the only remainings are the graphics, prototypes and some lines of code about the prediction of train-direction.

---------------------------------------

# Links
## Git Hub
https://github.com/alexaulbach/RadarTracker

## Factorio Forum
https://forums.factorio.com/viewtopic.php?f=97&t=35805

---------------------------------------

#Ideas/TODOs
V0.3:
- improve performance: reduce rescan, when train-state-change
- configurable scanned_radius for each entity-type.
- don't track back train when moving
- Add LICENSE.
- Playing for balancing.
- Make code nice. More clear modules...

V0.4:
- Take tracker-code out of control.lua and decrease complexity (nesting).
- Tracker should not calculate entity-status (managers job).
- Debugger should calculate output-string itself for speed reasons.
- The whole debugger functionality is questionable.
- Remote function for adding/deleting entities to/from list. Own file.
- Track only front entity if train is running.
- Don't track back entity if no reverse loco.
- Look if _new chunk_ would be found and only then chart that chunk.

V0.x:
- Tracker can scan only vehicles in range (I think to 2500 tiles radius, which is quite big). Big change: This needs to introduce to bind entities to trackers.
- The more trains a tracker needs to scan then, the slower it will refresh. Or the more energy it takes?
- Rotational-Tracker: Add a new radar-functionality, which scans 360˚ in a minute, but only those chunks, that have entities built on ground. Range depends only on power it can get.
- Once-Tracker: Scans when entity is built/removed
- Random-Tracker: Randomly scan minerals by revealing a similar "virtual" surface. Good to search for far away minerals without revealing current map.
- Spy tracker: Let's reveal the other side. If other force comes too near it gets the chance to be "marked" and you can see it for some while even when not longer in range.
- A tool or so to add entities to some of the trackers.
- With more trackers: spread their scan-times so, that they do not chart all at once.
- Upgrade function to convert (several) the train-tracker radars to to radar-tracker (good learning subproject).
- Scanner-robots. You can send them out and they will automatically reveal map.
- "Train Radar"  - some special type of locomotive/wagon. The train can be programmed to drive in circles to scan the area beside your rails in regular intervals, but needs excessive amounts of energy.


Changelog

#Version history
0.3.5 2016-12-12
- Fixed log-spam.
- Rewrote debugger and interface-code. Interface can now be configured module-wise.

0.3.4 2016-11-21
- New Icons for other trackers, thanks to YuokiTani.

0.3.3 2016-11-21
- New Icon from movement tracker, thanks to YuokiTani.
- Fix small bug with initialization of debugger.

0.3.2 2016-11-06
- Fixed: Debug-mode was initially on.
- Change: Reversed research-order of movement- and immobile-tracker and other slight changes with research. Reversed graphics, too.

0.3.1 2016-11-05
- Added release script.
- Second release on mods.factorio.com

0.3.0 2016-10-31
- Stabilized, immoveables and movement tracker now works reliable.
- Debug-mode. I think the idea of the "debugger" is cool. It helped me a lot to find the bugs, to see, when an entity changed state.
- New (ugly!) graphics to distinct better between the two types.

0.2.0 2016-10-10
- Silently reworking completely. No release.

0.1.3 2016-09-30 
- Introduced tracker types. Types are: trains (locos or wagons), stops (all train stops), cars (cars, tanks, satellite-uplink...); currently still only trains works.
- Rewrote initialization. on a very large map (8 GiB) with >300 trains initialisation takes only 20 seconds vs. hours from before.
- Avoid double reinitialization.
- Re-increased scan rate cause I want to reduce average number of scans per train. Also looks much better.

0.1.2 2016-09-04
- Initial version


control.lua
require 'stdlib.data.data'
require 'stdlib.area.area'
require 'stdlib.surface'

-- debug-purposes
inspect = require('inspect')

require 'config'
require 'init'
require 'container'
require 'manager'
require 'dbg'


RTDEF = {
	tracker = {
		none    = 1,    -- no scan
		running = 2,    -- movement-tracker   ; vehicles with speed ~= 0, nearly every 0.5 seconds
		waiting = 3,    -- wait-tracker   ; vehicles with speed == 0, every 10 seconds
		unmoved = 4,    -- immoveable-tracker ; static stuff, every 30 seconds
		rotating= 5,    -- rotational-tracker ; chunks that should be scanned rotational, every minute
		once    = 6,    -- scanning once like some kind of rocket or when a robot places/removes an entity
		random  = 7,    -- scanning randomly, like construction bots?
	},
	managers = {
		["train-stop"]  = "stops",
		["cargo-wagon"] = "trains",
		["locomotive"]  = "trains",
		["car"]         = "cars",
	}
}

stateComp = {
	[RTDEF.tracker.running] = { defines.train_state.on_the_path, defines.train_state.arrive_signal,
		defines.train_state.arrive_station, defines.train_state.manual_control,
		defines.train_state.stop_for_auto_control
	},

	[RTDEF.tracker.waiting] = { defines.train_state.wait_signal, defines.train_state.wait_station,
		defines.train_state.path_lost, defines.train_state.no_schedule,  defines.train_state.no_path,
		defines.train_state.manual_control_stop
	}
}
---------------------------------------------------------------------------------------------------

script.on_load(function()
	log("[RT] ----------------- on_load: do nothing--------------------------")
end)

script.on_init(function()
	log("[RT] ----------------- on_init--------------------------")
	Init.init()
end)

script.on_configuration_changed(function(event)
	log("[RT] ----------------- on_configuration_changed--------------------------")
	Init.init()
end)

script.on_event(defines.events.on_force_created, function(event)
	log("[RT] ----------------- on_force_created--------------------------")
	Init.clearInitialization()
	Init.init()
end)

script.on_event(defines.events.on_built_entity, function(event)
	local entity = event.created_entity
	local mngr = RTDEF.managers[entity.type]
	if mngr then
		log("[RT] built entity: " .. entity.unit_number .. " manager " .. mngr)
		manager.add(entity, mngr)
	end
end)

--[[
/c game.player.print("FRONT STOCK" .. game.player.selected.train.front_stock.unit_number);
game.player.print("BACK STOCK" ..game.player.selected.train.back_stock.unit_number);
for i, loco in pairs(game.player.selected.train.locomotives.front_movers) do
	game.player.print("Front Mover " .. i .. " . " .. loco.unit_number)
end
for i, loco in pairs(game.player.selected.train.locomotives.back_movers) do
	game.player.print("Back Mover " .. i .. " . " .. loco.unit_number)
end
for i, carr in pairs(game.player.selected.train.carriages) do
	game.player.print("Carr " .. i .. " . " .. carr.unit_number)
end

]]

script.on_event(defines.events.on_sector_scanned, function(event)
	local radar = event.radar
	if radar.name == "movement-tracker" then
		
		local trackers = { RTDEF.tracker.running }
		global.movement_tracker_count = global.movement_tracker_count + 1
		if global.movement_tracker_count % (_config["movement-tracker"].waiting_tracker_interval) == 0 then
			table.insert(trackers, RTDEF.tracker.waiting)
		end

		for _,tracker in pairs(trackers) do
			for unit_number, ntt in pairs(container.get_all_tracker_by_force(tracker, radar.force.name)) do
				local ent = ntt.entity
				if ent.valid then
					local area
					if ntt.manager == "cars" then
						-- turn into wait-mode if passenger has left
						if not ent.passenger and ntt.tracker ~= RTDEF.tracker.waiting then
							ntt.tracker = RTDEF.tracker.waiting
							container.set(ntt)
						end

						-- todo: cars should "forsee" one chunk more, than the players radar  --> config
						area = Area.adjust({
							{
								ent.position.x,
								ent.position.y
							}, {
								ent.position.x + ent.speed * _config["movement-tracker"].precognition * 10 * math.sin(2*math.pi * ent.orientation),
								ent.position.y - ent.speed * _config["movement-tracker"].precognition * 10 * math.cos(2*math.pi * ent.orientation)
							}
						})
	
					elseif ntt.manager == "trains" then
						area = Area.adjust({
							{
								ent.position.x,
								ent.position.y
							}, {
								ent.position.x + ent.train.speed * _config["movement-tracker"].precognition * math.sin(2*math.pi * ent.orientation),
								ent.position.y - ent.train.speed * _config["movement-tracker"].precognition * math.cos(2*math.pi * ent.orientation)
							}
						})
					end
					
					-- TODO: do not chart this chunk, if already charted (test if useful? How long does it take to chart a chunk?)
					area = Area.expand(area, _config["movement-tracker"].scanned_radius)
					radar.force.chart(radar.surface, area)
					dbg.charting(radar.surface, area, unit_number)
					
				else
					container.remove(ntt)
				end
			end
		end
	elseif radar.name == "immoveables-tracker" then
		for unit_number, ntt in pairs(container.get_all_tracker_by_force(RTDEF.tracker.unmoved, radar.force.name)) do
			local ent = ntt.entity
			if ent.valid then

				local x1 = ent.position.x
				local y1 = ent.position.y
				local area = Area.expand({{x1, y1}, {x1, y1}}, _config["immoveables-tracker"].scanned_radius)
				radar.force.chart(radar.surface, area)
				dbg.charting(radar.surface, area, unit_number)

				-- look if suddenly moving (again)
				if ntt.manager == "cars" then
					dbg.ftext(radar.surface, ent.position, "CAR " .. ntt.tracker .. " - " .. ent.speed)
					if ent.speed ~= 0.0 then
						ntt.tracker = RTDEF.tracker.running
						container.set(ntt)
					end
				elseif ntt.manager == "trains" then
					if ent.train.speed ~= 0.0 then
						ntt.tracker = RTDEF.tracker.running
						container.set(ntt)
					end
				end
				
			else
				container.remove(ntt)
			end
		end
	elseif radar.name == "rotational-tracker" then
		for unit_number, ntt in pairs(container.get_all_tracker_by_force(RTDEF.tracker.rotating, radar.force.name)) do
			log("[RT] rotational; " .. unit_number .. " ntt " .. ntt.entity.name)
		end
	end
end)


script.on_event(defines.events.on_train_changed_state, function(event)
	local train = event.train
	manager.add(train.front_stock, RTDEF.managers.locomotive)
end)

script.on_event(defines.events.on_player_driving_changed_state, function(event)
	local player = game.players[event.player_index]
	if not player.vehicle then
		-- not sitting in a vehicle...
		return
	end
	-- todo: need to watch for force?? Mabe in that case the entity is added to the own force-radars instead of vehicles?
	manager.add(player.vehicle)
end)


remote.add_interface("tr",
	{
		help = function()
			game.player.print("-----  RadarTracker: Remote functions  -----")
			game.player.print("|  help()  - This help")
			game.player.print("|  list()  - Show all currently tracked objects (ordered by force and tracker)")
			game.player.print("|  show()  - Show tracker entry of hovered entity (Only if tracked!)")
			game.player.print("|  init()  - (Re-)initialize everything")
			game.player.print("|  debug() - switch debug-functionality on/off")
			game.player.print("")
		end,
		
		list = function()
			for fn, trackers in pairs(global._ntttrkr) do
				prnt_n_log("--------- " .. fn .. " ---------")
				for tracker, units in pairs(trackers) do
					prnt_n_log("---- " .. tracker)
					for unit_number, ntt in pairs(units) do
						prnt_ntt(ntt, unit_number, fn) 
					end
				end
			end
		end, 
		
		show = function()
			entity = game.player.selected
			local unit_number
			if entity.type == 'train' then
				unit_number = entity.train.back_stock.unit_number
			else
				unit_number = entity.unit_number
			end
			local ntt = container.get(unit_number)
			if ntt then
				prnt_ntt(ntt, unit_number)
				if ntt.entity.unit_number ~= unit_number then
					prnt_n_log("UNIT NUMBER NOT EQUAL!")
				end
			end
		end,

		init = function()
			Init.clearInitialization()
			Init.init()
		end,
		
		debug = function()
			global.debugger = not global.debugger
			__switchDebug()
			game.player.print("[RT] Debugging: " .. inspect(global.debugger))
		end,
	}
)
